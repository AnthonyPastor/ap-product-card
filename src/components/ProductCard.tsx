import React, { createContext, useCallback } from 'react';

import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  onChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement[] | ReactElement ;
  children?: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  countValue?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  countValue,
  initialValues,
}: Props) => {
  const { counter, increaseBy, decreaseBy, maxCount, reset } = useProduct({
    product,
    onChange,
    countValue,
    initialValues,
  });

  const isMaxReached = useCallback(() => {
    return counter === maxCount;
  }, [counter, maxCount]);

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        decreaseBy,
        product,
        maxCount,
        isMaxReached: isMaxReached(),
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children &&
          children({
            count: counter,
            isMaxCountReached: isMaxReached(),
            increaseBy,
            decreaseBy,
            product,
            maxCount,
            reset,
          })}
      </div>
    </Provider>
  );
};
