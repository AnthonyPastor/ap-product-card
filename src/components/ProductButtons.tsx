import React, { useContext } from 'react';

import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
  style?: React.CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
  const { increaseBy, decreaseBy, counter, isMaxReached } = useContext(
    ProductContext
  );

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button className={styles.buttonMinus} onClick={() => decreaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        className={`${styles.buttonAdd} ${isMaxReached && styles.disabled}`}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  );
};
