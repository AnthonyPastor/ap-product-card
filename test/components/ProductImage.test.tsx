import React from 'react';
import renderer from 'react-test-renderer';

import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../data/products';

describe('pruebas en <ProductImage/>', () => {
  test('debe de mostrar el componenwte con la imagen personalizada', () => {
    const wrapper = renderer.create(<ProductImage img="https://hola.jpg" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe de mostrar el componente con la imagen de producto 2', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
