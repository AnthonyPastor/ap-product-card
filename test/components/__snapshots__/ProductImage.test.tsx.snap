// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <ProductImage/> debe de mostrar el componente con la imagen de producto 2 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Coffe Mug"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`pruebas en <ProductImage/> debe de mostrar el componenwte con la imagen personalizada 1`] = `
<img
  alt="Coffe Mug"
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;
